crates/build/src/build.rs

```diff
@@ -206,3 +211,132 @@ fn print_elf_paths_cargo_directives(target_elf_paths: &[(String, Utf8PathBuf)])
         println!("cargo:rustc-env=SP1_ELF_{}={}", target_name, elf_path);
     }
 }
+
+/// Verify that the locked version of `sp1-zkvm` in the Cargo.lock file is compatible with the
+/// current version of this crate.
+///
+/// This also checks to ensure that `sp1-sdk` is also the correct version.
+///
+/// ## Note: This function assumes that version compatibility is given by matching major and minor
+/// semver.
+///
+/// This is also correct if future releases sharing the workspace version, which should be the case.
+///
+/// # Errors:
+/// - If the locked version of `sp1-zkvm` is incompatible with the current toolchain version.
+/// - If the locked version of `sp1-sdk` is incompatible with the current toolchain version.
+/// -
+///
+/// # Panics
+/// - If we can't IO
+/// - If we can't find your lockfile
+/// - If we can't parse the version from the `cargo-prove`
+fn verify_locked_version(program_dir: impl AsRef<Path>, verbose: bool) -> Result<()> {
+    if std::env::var("DEV_SP1_SKIP_VERSION_CHECK").is_ok() {
+        println!("cargo:warning=Skipping version check");
+        return Ok(());
+    }
+
+    // We need an exception for our test fixtures.
+    if std::env::var("CARGO_PKG_NAME").unwrap_or_default() == "test-artifacts" {
+        return Ok(());
+    }
+
+    // This might be a workspace, need to optionally search parent dirs for lock files.
+    let canon =
+        program_dir.as_ref().canonicalize().expect("Failed to canonicalize path to program dir");
+
+    let mut lock_path = canon.join("Cargo.lock");
+    if !lock_path.is_file() {
+        let mut curr_path: &Path = canon.as_ref();
+
+        while let Some(parent) = curr_path.parent() {
+            let maybe_lock_path = parent.join("Cargo.lock");
+
+            if maybe_lock_path.is_file() {
+                lock_path = maybe_lock_path;
+                break;
+            } else {
+                curr_path = parent;
+            }
+        }
+
+        if !lock_path.is_file() {
+            panic!("Failed to find Cargo.lock in the program directory or its parent directories");
+        }
+    }
```

Comment:
`cargo-metadata` is a crate which should expose a method to get the `workspace_root`.